From b815ee752ea4e4a73b4fe091d1344d0d68e91c11 Mon Sep 17 00:00:00 2001
From: Pontus Fuchs <pontus.fuchs@gmail.com>
Date: Sun, 14 Oct 2012 15:35:06 +0200
Subject: [PATCH 77/82] Work around for mcast and bcast filters

For some reason the mcast and bcast filter flags does
not pass RX frames. I have no idea why it's not why :(
As a work around set the promiscuous flags.
---
 ar5523.c |   10 ++++++----
 1 file changed, 6 insertions(+), 4 deletions(-)

--- a/drivers/net/wireless/ath/ar5523/ar5523.c
+++ b/drivers/net/wireless/ath/ar5523/ar5523.c
@@ -1361,11 +1361,13 @@ static void ar5523_configure_filter(stru
 
 	*total_flags &= AR5523_SUPPORTED_FILTERS;
 
-	if (*total_flags & FIF_PROMISC_IN_BSS || *total_flags & FIF_OTHER_BSS)
-		filter |= UATH_FILTER_RX_PROM;
-
+	/* The filters seems strange. UATH_FILTER_RX_BCAST and
+	 * UATH_FILTER_RX_MCAST does not result in those frames being RXed.
+	 * The only way I have found to get [mb]cast frames seems to be
+	 * to set UATH_FILTER_RX_PROM. */
 	filter |= UATH_FILTER_RX_UCAST | UATH_FILTER_RX_MCAST |
-		  UATH_FILTER_RX_BCAST | UATH_FILTER_RX_BEACON;
+		  UATH_FILTER_RX_BCAST | UATH_FILTER_RX_BEACON |
+		  UATH_FILTER_RX_PROM;
 
 	ar5523_set_rxfilter(ar, 0, UATH_FILTER_OP_INIT);
 	ar5523_set_rxfilter(ar, filter, UATH_FILTER_OP_SET);
